lisa/nic.py

```diff
@@ -0,0 +1,222 @@
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT license.
+
+import ipaddress
+import itertools
+import re
+from typing import Dict, List
+
+from assertpy import assert_that
+
+from lisa import Node
+
+
+class NicInfo:
+
+    # Class for info about an single upper/lower nic pair.
+    # devices using SRIOV on azure typically have an upper synthetic device
+    # paired with a lower SRIOV Virtual Function (VF) device that
+    # enables the passthrough to the physical NIC.
+
+    def __init__(
+        self,
+        upper: str,
+        lower: str = "",
+        pci_slot: str = "",
+    ) -> None:
+        self._has_secondary = lower != "" or pci_slot == ""
+        self._upper = upper
+        self._lower = lower
+        self._pci_slot = pci_slot
+        self._ip_addr = ""
+        self._mac_addr = ""
+
+    def has_secondary(self) -> bool:
+        return self._has_secondary
+
+    def __str__(self) -> str:
+        return (
+            "NicInfo:\n"
+            f"upper: {self._upper}\n"
+            f"lower: {self._lower}\n"
+            f"pci_slot: {self._pci_slot}\n"
+            f"ip_addr: {self._ip_addr}\n"
+            f"mac_addr: {self._mac_addr}\n"
+        )
+
+    def testpmd_include(self, vdev_type: str) -> str:
+        assert_that(self._has_secondary).is_true().described_as(
+            (
+                f"This interface {self._upper} does not have a lower interface "
+                "and pci slot associated with it. Aborting."
+            )
+        )
+        return f'--vdev="{vdev_type},iface={self._upper}" --allow "{self._pci_slot}"'
+
+
+class NodeNics:
+
+    # Class for all of the nics on a node. Contains multiple NodeNic classes.
+    # Init identifies upper/lower paired devices and the pci slot info for the lower.
+
+    def _get_nic_names(self) -> List[str]:
+        # identify all of the nics on the device, excluding tunnels and loopbacks etc.
+        result = self._node.execute(
+            " ls /sys/class/net/ | grep -Ev $(ls /sys/devices/virtual/net)",
+            shell=True,
+            sudo=True,
+        )
+        nic_names = result.stdout.splitlines()
+        for item in nic_names:
+            assert_that(item).described_as(
+                "nic name could not be found"
+            ).is_not_equal_to("")
+        return nic_names
+
+    def _get_nic_device(self, nic_name: str) -> str:
+        slot_info_result = self._node.execute(
+            f"readlink /sys/class/net/{nic_name}/device"
+        )
+        slot_info_result.assert_exit_code()
+        base_device_result = self._node.execute(f"basename {slot_info_result.stdout}")
+        base_device_result.assert_exit_code()
+        # todo check addr matches expectation
+        return base_device_result.stdout
+
+    def _get_node_nic_info(self, nic_list: List[str]) -> None:
```

Comment:
1. It doesn't need to pass in a couple of names, those names can be queried in this method.
2. If it doesn't return a value, call it like `_initialize`. It can inherit from `InitializableMixin`. If it calls `get_...`, let it return the dict of nics.